modules/lsp.nix

```
@@ -0,0 +1,96 @@
+{
+  lib,
+  config,
+  ...
+}:
+let
+  inherit (lib) mkOption types;
+
+  cfg = config.lsp;
+
+  serverType = types.submodule {
+    options = {
+      enable = lib.mkEnableOption "the language server";
+
+      config = mkOption {
+        type = with types; attrsOf anything;
+        description = ''
+          Configurations for each language server.
+        '';
+        default = { };
+        example = {
+          cmd = [
+            "clangd"
+            "--background-index"
+          ];
+          root_markers = [
+            "compile_commands.json"
+            "compile_flags.txt"
+          ];
+          filetypes = [
+            "c"
+            "cpp"
+          ];
+        };
+      };
+    };
+  };
+in
+{
+  options.lsp = {
+    servers = mkOption {
+      type = types.attrsOf serverType;
+      description = ''
+        Wrapper of the `vim.lsp` API.
+      '';
+      default = { };
+      example = {
+        "*".config = {
+          root_markers = [ ".git" ];
+          capabilities.textDocument.semanticTokens = {
+            multilineTokenSupport = true;
+          };
+        };
+        luals.enable = true;
+        clangd = {
+          enable = true;
+          config = {
+            cmd = [
+              "clangd"
+              "--background-index"
+            ];
+            root_markers = [
+              "compile_commands.json"
+              "compile_flags.txt"
+            ];
+            filetypes = [
+              "c"
+              "cpp"
+            ];
+          };
+        };
+      };
+    };
+  };
+
+  config = {
+    extraConfigLua =
+      let
+        serversToEnable = lib.pipe cfg.servers [
+          (lib.filterAttrs (_: props: props.enable))
+          (lib.mapAttrsToList (serverName: _: serverName))
+        ];
+      in
+      ''
+        do
+          local __nixvimLspServersConfigs = ${lib.nixvim.toLuaObject cfg.servers}
+
+          for server_name, server_props in ipairs(__nixvimLspServersConfigs) do
```

```suggestion
          for server_name, server_props in pairs(__nixvimLspServersConfigs) do
```
this fixes the config being passed for me

---

But seems to barf when i don't have a config... lol